### PR TITLE
feat(ir): Top-down retargeter for MemoryReuse accumulator chains

### PR DESCRIPTION
**Pull request by Hzfengsy**
## Summary

- Add a Step-0 **top-down retargeter** to `MemoryReuse` that rewrites each `ForStmt`'s yield-chain producer to land directly in the iter_arg's MemRef (eliminates `tile.move` insertions for accumulator chains).
- Tighten the **InitMemRef** invariant so `return_vars` share the *iter_arg*'s MemRef (not the yield value's), giving the retargeter a consistent `(initValue, iter_arg, return_var)` target.
- Update 8 existing tests whose expectations described the old `tile.move`-insertion pattern; add 2 new tests for the retargeter's positive (Acc `matmul_acc` chain) and negative (target-still-live) paths.

## Motivation

For `Acc` → `Acc` accumulator chains (`matmul → matmul_acc → matmul_acc …`) the old lifetime-coloring step reconciled a yield/iter_arg MemRef mismatch by inserting a `tile.move` at the yield. An `Acc → Acc` `tile.move` lowers to a `pto.tmov` on `Acc` memory that the codegen backend rejects, so otherwise-legal accumulator loops failed to compile end-to-end (follow-up to the work in #1083).

## Algorithm

On entry to `MemoryReuse`, a new Step-0 walks each `ForStmt` top-down. For every `iter_arg → yield-value` pair it takes the iter_arg's MemRef as the canonical "target" and tries to retype the yield producer directly onto that MemRef.

- **Unconstrained producers** (plain `add`, `move`, `matmul`, …): retyped to write directly into the target MemRef, provided the target's base `Ptr` is not read between the producer and the yield in the same branch body (liveness check).
- **Pinned producers** (ops with `set_output_reuses_input`, e.g. `matmul_acc`): the output MemRef cannot be changed, so the retargeter recurses onto the pinned input (typically the iter_arg itself) and only plans the LHS retype once the upstream placement is feasible.
- **IfStmt return_vars**: recurse into both branches' yield values with the same target; retype the return_var when both branches succeed.

When the retargeter succeeds, the producer lands directly in the loop-carry buffer and the downstream `YieldFixup` pass has nothing to reconcile, so no `tile.move` ops are emitted. When it declines (e.g. target still live after the producer), `YieldFixup` handles the mismatch exactly as before.

### InitMemRef invariant the retargeter depends on

The retargeter assumes `initValue`, `iter_arg`, and `return_var` all share a single MemRef for the whole loop. `InitMemRef` previously patched `return_vars` from the *yield value*'s MemRef, which left the iter_arg out of sync whenever the yield producer lived in a different buffer. This PR patches `return_vars` from the *iter_arg*'s MemRef instead, so the retargeter can consume a consistent target.

### Complexity note

The retargeter runs a single O(N) IR walk to build the def map, then one `TryRetargetVar` per iter_arg plus recursion into `IfStmt` branches. The liveness check scans the tail of the `AssignStmt`'s containing body; in the worst case of a deep linear chain this is O(N²). In practice body tails are short (matmul/accumulator loops have a handful of producers each), so the realised cost is well below that bound. This super-linear worst case is documented in a comment at the top of the retargeter section per the project's `pass-complexity.md` rule.

## Test expectation updates — per-test justification

Each of the 8 updated tests had Expected IR describing the *old* `producer-on-fresh-buffer + tile.move-at-yield` pattern. Every update swaps that pattern for the retargeter's direct-retype behaviour. No test's *semantic* intent is relaxed — only the tactical buffer-allocation detail changes.

| Test | Old expected | New expected | Why the new expectation is correct |
| ---- | ------------ | ------------ | ---------------------------------- |
| `TestYieldFixup::test_producer_retyped_to_iter_arg_buffer` (was `test_tile_move_inserted_when_memrefs_diverge`) | `next_0` on fresh `mem_vec_3`; `tile.move(next_0) → mem_vec_2` | `next_0` directly on `mem_vec_2`; `extra_0` keeps `mem_vec_3` | Loop-carry unified with iter_arg — exactly what the test name promised, achieved without the intermediate move |
| `TestYieldFixup::test_simple_loop_memrefs_unified` | `next_0` on `mem_vec_3`; `tile.move` to `mem_vec_2` | `next_0` directly on `mem_vec_2`; `mem_vec_3` dropped | Loop-carry unified — stricter: one buffer total |
| `TestYieldFixup::test_multiple_iter_args_producers_retyped_independently` (was `…_partial_mismatch`) | Two `tile.move`s unifying `next_0/next_1` back to iter buffers | `next_0` on `mem_vec_2`, `next_1` on `mem_vec_3`; `extra_*` share `mem_vec_4` | Each iter_arg gets its own producer placed directly in its buffer |
| `TestControlFlow::test_var_used_in_nested_if_shares_iter_arg_buffer` (was `…_not_reused_in_loop`) | `tile_c` on fresh `mem_vec_3`; `tile.move(if_result) → mem_vec_2` | `tile_c`, `if_result`, `loop_out` all on `mem_vec_2` | `tile_a`/`acc` already share `mem_vec_2` via `InitMemRef`; retargeter collapses the yield chain |
| `TestControlFlow::test_loop_local_var_can_be_reused` | `tile_x/y/z` share `mem_vec_3`; `tile.move(tile_z) → mem_vec_2` | `tile_x/y` share `mem_vec_3`; `tile_z` directly on `mem_vec_2` | Loop-locals can still share their own scratch; the yield producer is retyped |
| `TestControlFlow::test_nested_for_loops_outer_var_extends_to_outer_end` | Moves at both inner and outer yields; 4 allocations | `tile_b` on inner iter buffer `mem_vec_4`; `tile_d` on outer iter buffer `mem_vec_3`; 3 allocations | Nested loop-carry unification at both levels |
| `TestControlFlow::test_for_with_if_multiple_vars_competing` | `tile_c/tile_d` on `mem_vec_5`; `tile.move(if_result) → mem_vec_4` | `tile_c/tile_d/if_result/loop_out` all on `mem_vec_4` | Cross-branch producer sharing via the iter buffer |
| `TestControlFlow::test_loop_return_var_blocks_init_memref_reuse` (#768 regression) | `acc_next` on `mem_vec_5`; move to `mem_vec_3` | `acc_next` directly on `mem_vec_3`; `resid` still on `mem_vec_5` (NOT `mem_vec_3`) | **The critical #768 invariant still holds:** `resid` does not clobber the loop-carry buffer. Only the now-unnecessary move is removed. |

## New tests

- `TestTopDownRetargeter::test_acc_chain_pinned_producer_shares_iter_arg_buffer` — a `matmul_acc` chain over `Acc` memory. Exercises the pinned-producer path: `matmul_acc`'s `set_output_reuses_input(0)` forces the retargeter to recurse through the pin to the iter_arg (already on target), then retype `acc_next`. Verifies that the full chain shares a single `mem_acc_*` allocation and that **no** `tile.move` ops remain in the loop body.
- `TestTopDownRetargeter::test_retargeter_declines_when_target_still_live` — a loop body where an op reads `acc_0` *after* the yield producer writes. Verifies the retargeter correctly declines to retype (leaves the producer on its own buffer), so that `YieldFixup` can insert the `tile.move` safely.

## Test plan

- [x] `cmake --build build --parallel` succeeds in the worktree
- [x] `tests/ut/ir/transforms/test_memory_reuse.py` (36 tests, includes the 8 updated + 2 new) — pass
- [x] `tests/ut/ir/transforms/test_init_memref.py` — pass
- [x] Full IR suite `tests/ut/ir/ -n auto` — 2706 passed, 13 skipped
- [x] `clang-tidy` clean on the diff
- [x] Pre-commit hooks pass (clang-format, cpplint, ruff, pyright)

### CODE REVIEW

File: src/ir/transforms/memory_reuse_pass.cpp [hunk 1]
```diff
@@ -76,6 +76,380 @@ class VarUseCollector : public IRVisitor {
...
+  }
+
+  StmtPtr VisitStmt_(const AssignStmtPtr& op) override {
+    auto rit = rewrites_.find(op->var_);
+    if (rit != rewrites_.end()) {
+      auto new_var = std::make_shared<Var>(op->var_->name_hint_, rit->second, op->var_->span_);
+      var_substitution_[op->var_] = new_var;
+    }
+    return IRMutator::VisitStmt_(op);
+  }
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/transforms/memory_reuse_pass.cpp:501]: This is a false positive. `IRMutator::VisitStmt_(AssignStmt)` at `src/ir/transforms/mutator.cpp:301-324` does visit `op->var_` via `ExprFunctor::VisitExpr` (line 304), which dispatches to this mutator's `VisitExpr_(const VarPtr&)` override and returns the substituted var. The base then rebuilds the AssignStmt with the new var at line 317-321. The pre-register-then-delegate pattern here is correct and is also used by other RetypeApplier-style mutators in the tree. No change needed.

File: src/ir/transforms/memory_reuse_pass.cpp [hunk 2]
```diff
@@ -76,6 +76,380 @@ class VarUseCollector : public IRVisitor {
...
+    // Pre-register substitutions for any retargeted return_var; the default
+    // IRMutator IfStmt handler will then pick them up through VisitExpr_(Var).
+    for (const auto& rv : op->return_vars_) {
+      auto rit = rewrites_.find(rv);
+      if (rit != rewrites_.end()) {
+        var_substitution_[rv] = std::make_shared<Var>(rv->name_hint_, rit->second, rv->span_);
+      }
+    }
+    return IRMutator::VisitStmt_(op);
+  }
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/transforms/memory_reuse_pass.cpp:513]: Same as above: `IRMutator::VisitStmt_(IfStmt)` at `src/ir/transforms/mutator.cpp:326-373` explicitly iterates `return_vars_` (lines 348-362) and visits each via `ExprFunctor::VisitExpr`, which hits this mutator's `VisitExpr_(VarPtr)` override. The pre-register-then-delegate approach is correct. No change needed.

File: src/ir/transforms/memory_reuse_pass.cpp [hunk 3]
```diff
@@ -76,6 +76,380 @@ class VarUseCollector : public IRVisitor {
...
+    // producers always have a sibling YieldStmt, so this branch only fires
+    // for AssignStmts the retargeter would never plan to rewrite in the
+    // first place; treating it as "dead" is both safe and unreachable-in-
+    // practice for the rewrite path.
+    if (!seq) return true;
+    auto it = std::find(seq->stmts_.begin(), seq->stmts_.end(), def.assign_stmt);
+    if (it == seq->stmts_.end()) return true;
+    for (++it; it != seq->stmts_.end(); ++it) {
+      if (BodyReadsBase(*it, target_base)) return false;
+    }
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/transforms/memory_reuse_pass.cpp]: Good catch — fixed in 6a088354. The liveness check now walks the AssignStmt's full ancestor chain (tracked during the DefMap build), scans sibling tails at every enclosing SeqStmts, and stops at the owning ForStmt. Added test_retargeter_declines_when_read_after_nested_if as a regression test for exactly this scenario (yield producer inside an IfStmt branch with a subsequent read of the iter_arg after the IfStmt).

File: src/ir/transforms/memory_reuse_pass.cpp [hunk 4]
```diff
@@ -76,6 +76,380 @@ class VarUseCollector : public IRVisitor {
...
+      // Reading iter_arg init = reading some outer var; count as read only if base matches.
+      for (const auto& ia : f->iter_args_) {
+        if (ia->initValue_ && CollectBasesUsed(ia->initValue_).count(target_base) > 0) return true;
+      }
+      if (BodyReadsBase(f->body_, target_base)) return true;
+      return false;
+    }
+    if (auto s = As<ScopeStmt>(stmt)) return BodyReadsBase(s->body_, target_base);
+    return false;
+  }
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/transforms/memory_reuse_pass.cpp]: Fixed in 6a088354. Replaced the hand-enumerated BodyReadsBase with a generic `SubtreeReadBaseCollector` (IRVisitor subclass) that walks the whole stmt subtree and collects MemRef base Ptrs for every TileType Var read, skipping only the LHS of AssignStmt. All stmt/expression kinds are now covered through the default IRVisitor dispatch, so new stmt types (e.g. EvalStmt) won't silently bypass the check.

File: src/ir/transforms/memory_reuse_pass.cpp [hunk 5]
```diff
@@ -76,6 +76,383 @@ class VarUseCollector : public IRVisitor {
...
+      d.kind = VarDef::kIfReturn;
+      d.control_stmt = op;
+      d.return_idx = i;
+      defs[rv] = d;
+    }
+    ancestor_stack_.push_back(op);
+    VisitStmt(op->then_body_);
+    if (op->else_body_.has_value()) VisitStmt(op->else_body_.value());
+    ancestor_stack_.pop_back();
+  }
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/transforms/memory_reuse_pass.cpp:177]: Fixed in ce87a027. Added `VisitStmt_` overrides for all five ScopeStmt subclasses (InCore / AutoInCore / Cluster / Hierarchy / Spmd) via a `VisitScope` helper that pushes the scope onto `ancestor_stack_` before descending into its body. The liveness walk now correctly finds its path-child when stepping from a scope body SeqStmts up through the scope into the enclosing loop body.

File: src/ir/transforms/memory_reuse_pass.cpp [hunk 6]
```diff
@@ -76,6 +76,383 @@ class VarUseCollector : public IRVisitor {
...
+    auto reuse_idx = reg.GetEntry(call->op_->name_).GetOutputReusesInputArg();
+    if (reuse_idx.has_value()) {
+      // Pinned output: can't change this stmt's LHS MemRef; recurse onto pinned input.
+      if (*reuse_idx >= call->args_.size()) return false;
+      auto input_var = As<Var>(call->args_[*reuse_idx]);
+      if (!input_var) return false;
+      if (!TryRetargetVar(input_var, target, target_memory)) return false;
+      // Also record that `var`'s MemRef should follow the pinned input to target.
+      PlanRewrite(var, target, target_memory);
+      return true;
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/transforms/memory_reuse_pass.cpp]: Fixed in ce87a027 by declining retargeting for view ops. Added `IsOutputMemoryInheritInput(entry)` helper that checks whether the op has a memory spec with no `output_reuses_input_arg` and a `deduce_output_memory` that returns nullopt — the signature left by `set_output_memory_inherit_input`. `RetargetAssign` now returns false early for these ops, so producer-view byte_offset / size are preserved.

File: src/ir/transforms/memory_reuse_pass.cpp [hunk 7]
```diff
@@ -76,6 +76,383 @@ class VarUseCollector : public IRVisitor {
...
+    }
+
+    // Unconstrained: check liveness, then plan retype.
+    if (!IsTargetDeadAtAssign(def, target->base_.get())) return false;
+    PlanRewrite(var, target, target_memory);
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/transforms/memory_reuse_pass.cpp:383]: Fixed in ce87a027. Added a `HasKwarg(call, "target_memory")` check in `RetargetAssign`: any call carrying an explicit `target_memory` kwarg (tile.move / tile.load / etc.) is declined, because rewriting only the LHS type would leave the kwarg out of sync. The retargeter now only rewrites ops whose output MemRef is fully determined by the LHS type alone.

File: tests/ut/ir/transforms/test_memory_reuse.py [hunk 8]
```diff
@@ -1304,22 +1297,16 @@ def main(
...
                         pl.tile.add(extra_1, acc_1)
                     )
-                    next_0_mv: pl.Tile[[64, 64], pl.FP32, pl.MemRef(mem_vec_2, 0, 16384), pl.Mem.Vec] = (
-                        pl.tile.move(next_0, target_memory=pl.Mem.Vec)
-                    )
-                    next_1_mv: pl.Tile[[64, 64], pl.FP32, pl.MemRef(mem_vec_3, 0, 16384), pl.Mem.Vec] = (
-                        pl.tile.move(next_1, target_memory=pl.Mem.Vec)
-                    )
-                    out_0, out_1 = pl.yield_(next_0_mv, next_1_mv)
+                    out_0, out_1 = pl.yield_(next_0, next_1)
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply tests/ut/ir/transforms/test_memory_reuse.py]: Fixed in ce87a027 — both occurrences of the unused `out_1` tuple element (Before and Expected programs of the multi-iter test) are now `_out_1`.

File: tests/ut/ir/transforms/test_memory_reuse.py [hunk 9]
```diff
@@ -2071,6 +2058,266 @@ def main(
...
+                        )
+                    side: pl.Tile[[64, 64], pl.FP32, pl.MemRef(mem_vec_3, 0, 16384), pl.Mem.Vec] = (
+                        pl.tile.mul(acc_0, acc_0)
+                    )
+                    _use: pl.Tensor[[64, 64], pl.FP32, pl.MemRef("mem_ddr_1", 0, 16384)] = pl.tile.store(
+                        side, [0, 0], output
+                    )
+                    if_result_mv: pl.Tile[[64, 64], pl.FP32, pl.MemRef(mem_vec_2, 0, 16384), pl.Mem.Vec] = (
+                        pl.tile.move(if_result, target_memory=pl.Mem.Vec)
+                    )
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply tests/ut/ir/transforms/test_memory_reuse.py:2310]: Good catch. The original test used `side = pl.mul(acc_0, acc_0)` + `pl.store(side, ...)` as the post-IfStmt read, and the reuse pass's lifetime analysis coalesced `side` onto the same buffer as `if_result`, producing Expected IR that looked like `side` would clobber `if_result` before the tile.move read it. That coalescing is a pre-existing characteristic of the lifetime analysis (not something this PR introduces), but it obscured what the test is actually asserting. Rewrote in ce87a027 to use `pl.store(acc_0, [0, 0], output)` directly as the post-IfStmt read — no intermediate Tile, no lifetime coalescing concern, same retargeter decline behaviour.

**Comment by Hzfengsy** [comment]
@codex review

**Review by lyfne123 — APPROVED** [review]
